Index: app/src/main/kotlin/util/EventQueue.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- app/src/main/kotlin/util/EventQueue.kt	(date 1566932159530)
+++ app/src/main/kotlin/util/EventQueue.kt	(date 1566932159530)
@@ -0,0 +1,26 @@
+package util
+
+import java.util.*
+
+class EventQueue<T> {
+
+    var handler: (T) -> Unit = {}
+
+    private val queue = LinkedList<T>()
+    private var processing = false
+
+    fun process(event: T) {
+        queue.push(event)
+        processQueue()
+    }
+
+    private tailrec fun processQueue() {
+        synchronized(this) {
+            if (processing || queue.isEmpty()) return
+            processing = true
+            handler(queue.poll())
+            processing = false
+        }
+        processQueue()
+    }
+}
\ No newline at end of file
